rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RemoteFetchProcessor.java

```diff
@@ -0,0 +1,114 @@
+import org.apache.hadoop.conf.Configuration;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.ExecutionException;
+
+public class RemoteFetchProcessor extends AbstractScheduledService
+{
+  private Configuration conf;
+  private Queue<FetchRequest> processQueue = null;
+  private FileDownloader downloader = null;
+
+  int processThreadInitalDelay;
+  int processThreadInterval;
+  long requestProcessDelay;
+
+  private static final Log log = LogFactory.getLog(RemoteFetchProcessor.class);
+
+  public RemoteFetchProcessor(Configuration conf)
+  {
+    this.conf = conf;
+    this.processQueue = new ConcurrentLinkedQueue<FetchRequest>();
+    this.downloader = new FileDownloader(conf);
+
+    this.processThreadInitalDelay = CacheConfig.getProcessThreadInitialDelayInMs(conf);
+    this.processThreadInterval = CacheConfig.getProcessThreadIntervalInMs(conf);
+    this.requestProcessDelay = CacheConfig.getRemoteFetchProcessIntervalInMS(conf);
+  }
+
+  public void addToProcessQueue(String remotePath, long offset, int length, long fileSize, long lastModified)
+  {
+    long requestedTime = System.currentTimeMillis();
+    FetchRequest request = new FetchRequest(remotePath, offset, length, fileSize, lastModified, requestedTime);
+    processQueue.add(request);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception
+  {
+    long currentTime = System.currentTimeMillis();
+
+    processRequest(currentTime);
+  }
+
+  protected void processRequest(long currentTime) throws IOException, InterruptedException, ExecutionException
+  {
+    ConcurrentMap<String, DownloadRequestContext> contextMap = mergeRequests(currentTime);
+    processRemoteFetchRequest(contextMap);
+
+    // After every iteration we are clearing the map
+    contextMap.clear();
+  }
+
+  protected ConcurrentMap<String, DownloadRequestContext> mergeRequests(long currentTime)
+  {
+    // Till the queue is not empty or there are no more requests which came in before the configured delay time
+    // we are going to collect the requests and process them
+    ConcurrentMap<String, DownloadRequestContext> contextMap = new ConcurrentHashMap<String, DownloadRequestContext>();
+    while (!processQueue.isEmpty()) {
+      FetchRequest request = processQueue.peek();
+      if (currentTime - request.getRequestedTime() < this.requestProcessDelay) {
+        break;
+      }
+
+      if (!contextMap.containsKey(request.getRemotePath())) {
+        DownloadRequestContext context = new DownloadRequestContext(request.getRemotePath(), request.getFileSize(),
+            request.getLastModified());
+        contextMap.putIfAbsent(request.getRemotePath(), context);
+      }
+      contextMap.get(request.getRemotePath()).addDownloadRange(request.getOffset(),
```

Comment:
This could cause corruption of data (probably for a short interval only) in case a file is modified at source and we have some FetchRequests for previous version and a few for new one. We should drop the DownloadRequest if we find a mismatch in modifiedTime across fetchRequests